dspace-api/src/main/java/org/dspace/app/suggestion/SuggestionServiceImpl.java

```diff
@@ -0,0 +1,191 @@
+ */
+package org.dspace.app.suggestion;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.annotation.Resource;
+
+import org.apache.logging.log4j.Logger;
+import org.dspace.core.Context;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SuggestionServiceImpl implements SuggestionService {
+    private static final Logger log = org.apache.logging.log4j.LogManager.getLogger(SuggestionServiceImpl.class);
+
+    @Resource(name = "suggestionProviders")
+    private Map<String, SuggestionProvider> providersMap;
+
+    @Override
+    public List<SuggestionProvider> getSuggestionProviders() {
+        if (providersMap != null) {
+            return providersMap.values().stream().collect(Collectors.toList());
+        }
+        return null;
+    }
+
+    @Override
+    public SuggestionTarget find(Context context, String source, UUID id) {
+        if (providersMap.containsKey(source)) {
+            return providersMap.get(source).findTarget(context, id);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public long countAll(Context context, String source) {
+        if (providersMap.containsKey(source)) {
+            return providersMap.get(source).countAllTargets(context);
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public List<SuggestionTarget> findAllTargets(Context context, String source, int pageSize, long offset) {
+        if (providersMap.containsKey(source)) {
+            return providersMap.get(source).findAllTargets(context, pageSize, offset);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public long countAllByTarget(Context context, UUID target) {
+        int count = 0;
+        for (String provider : providersMap.keySet()) {
+            if (providersMap.get(provider).countUnprocessedSuggestionByTarget(context, target) > 0) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    @Override
+    public List<SuggestionTarget> findByTarget(Context context, UUID target, int pageSize, long offset) {
+        List<SuggestionTarget> fullSourceTargets = new ArrayList<SuggestionTarget>();
+        for (String source : providersMap.keySet()) {
+            SuggestionTarget sTarget = providersMap.get(source).findTarget(context, target);
+            if (sTarget != null && sTarget.getTotal() > 0) {
+                fullSourceTargets.add(sTarget);
+            }
+        }
+        fullSourceTargets.sort(new Comparator<SuggestionTarget>() {
+            @Override
+            public int compare(SuggestionTarget arg0, SuggestionTarget arg1) {
+                return -(arg0.getTotal() - arg1.getTotal());
+            }
+        }
+        );
+        return fullSourceTargets.stream().skip(offset).limit(pageSize).collect(Collectors.toList());
```

Comment:
Because an Item can be targeted by multiple suggestions with different sources. For this reason, firstly we collect the suggestions for each managed source, then we sort and page the list manually.

Reply:
@frabacche :  I'm concerned with this approach because we've hit performance issues with doing sort/pagination manually in the past.  As I noted, what if `fullSourceTargets` is very large (e.g. 200 SuggestionTarget objects), but you only need the first page (of 5)? 

In that scenario, you first must load & sort all 200 objects into memory even if you only need to display the first 5.  Then, when you click on the next page, you have to load and sort all 200 objects **again** to get the next 5.   Essentially pagination will work extremely slowly because you always load every object into memory for every page.

This is the behavior we've seen in the past that caused major performance problems. See for example #9078 which had to remove similar manual pagination on our Group & EPerson endpoints as those endpoints performed extremely slow pagination.

To me, this sort of manual pagination implies a possible design issue.  We shouldn't need to load hundred (or more) of objects into memory and sort them if all we want is the first five.

Reply:
Consider that [**SuggestionTarget**](https://github.com/DSpace/DSpace/blob/c83e3fb96f0fcb23f529e3e6de984f962af3bb34/dspace-api/src/main/java/org/dspace/app/suggestion/SuggestionTarget.java) object is just a triple of _[source, targetid, total]_ . The solrQuery written at [SolrSuggestionStorageServiceImpl] (https://github.com/DSpace/DSpace/blob/c83e3fb96f0fcb23f529e3e6de984f962af3bb34/dspace-api/src/main/java/org/dspace/app/suggestion/SolrSuggestionStorageServiceImpl.java#L282-L301)
produces the SuggestionTarget result:
_source_ is a string from the keyset of the hashmap defined on spring context named _suggestionProviders_ see ;
_targetid_ is the uuid of the researcher;
_total_ is the counter of the solr query, searching for specified source, specified target, processed=false;

Let's say the source _openaire_ has 200 suggestion records for target _Bacchelli_: the solr query executed runtime returns one record: [openaire, Bacchelli, 200]. Even if many sources are managed - we are going to have one suggestion target result for each targetid. 

Eventually at the moment we have one single managed source - more are to come - but not so many more to cause performance issue: even if we are paginating this way.

Reply:
@frabacche : The issue here is that SuggestionTarget is also storing entire Item objects (see `setTarget()` method and `findItem` in the code you linked above).  Item objects themselves can be rather large objects.  Loading 200 or more Item objects into memory at once can have an impact on performance.

So, SuggestionTarget is not just merely storing strings/UUIDs.  It has references to Item objects, and that is the primary concern here.

Reply:
@frabacche: To clarify my prior comment further.  If we cannot fix the pagination problem (of loading all SuggestionTargets into memory to paginate them), another possible solution would be to change SuggestionTarget to **only store the Item UUID**.  

But, currently SuggestionTarget has an internal reference to the entire `Item` object & the code above has to do a lookup using `ItemService.find(context, UUID)`.  This lookup & keeping all Items in memory at once is the main issue.  That can be expensive for performance when we are talking about looking up & storing hundreds or thousands of Items.

Reply:
hi @tdonohue we have reviewed your feedback and I agree with you that an improvement is required here.
The initial design delegate the responsibility of the storage of the suggestion to the individual provider this make impossible to paginate when multiple providers are involved (i.e. in a findAllTarget). We will check if it is possible to move this responsibility to the service with a reasonable effort otherwise we will proceed as by your suggestion to only deal with the uuid until the very last moment where, once that the results are restricted to the specific page, we will embed also the item. We will update you on that in 2-3 days

Reply:
Thanks @abollini .  That sounds good to me.

Reply:
sorry, I just noted that I sent the wrong version of the message that was draft this morning before to finalize the analysis...
it is true that there is a potential area of improvement if we move the responsibility of the storage out of the single suggestion provider but this would not have any visible impact. 
The method is just retrieving from each source (we have one "openaire" in this PR but we will expect other 3 from other PRs "coar notify", "dspace users feedback", "orcid webhook") the aggregated number of suggestions for the same target.
So, our "full list" would consist of 4 SuggestionTarget all related to the same target (Item) and also in future is really unlikely that institution will have a so large number of additional / custom sources to really trigger pagination.
In short, the current design inefficiency is that instead than performing a single solr query by target uuid and looking to the value of a facet on the "sources" we are performing a separate solr query for each source filtering by source and target uuid and collecting the result count for each source (so less than 4 if all the providers should be merged and enabled for a given repository),  
According to that we suggest to keep the code as is

Reply:
@abollini : In that case, I request that we better describe `SuggestionTarget` or add better comments to this block of code.

It appears I've been misunderstanding that a `SuggestionTarget` might be created *per Person Item*  (which is why I assumed that if you have 200 items, you'd have 200 SuggestionTargets).  Instead, it sounds like we have a single `SuggestionTarget` **per external Source**.  So,  even if we have 200 Items, it doesn't matter because we will have a smaller limit of external sources.

So, please add a comment either in this `SuggestionServiceImpl.java` or in `SuggestionTarget` class itself to explain that these Targets are created per Source.

This also brings up a followup question.  Why, then, should we store the **same Item** in every SuggestionTarget?  Shouldn't we still update this to just store the **Item UUID**?  That way we only lookup the Item *once* instead of 4 times?

I think I'm better understanding the design...I'm just pointing out that both *you* and I were confused here.  That implies we need to document this design better in code comments.  Even just adding a note above this manual pagination that says something like:
```
// We perform manual pagination here because there is a limited number of SuggestionTargets
// Each external source will only result in one SuggestionTarget. So even with 10 sources, we only will have 10 SuggestionTargets at a maximum.
```

Reply:
I have added the javadoc and updated the code in one place where I found that was indeed possible to introduce real pagination.
Looking multiple time to the same item is not a real issue as it is done using the PK within the same hibernate session (i.e. the exact same java object will be shared across all the instance of the SuggestionTarget)